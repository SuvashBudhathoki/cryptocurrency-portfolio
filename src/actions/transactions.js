import database from "../firebase/firebase";

//Add_Transaction h

export const addTransaction = (transaction) => ({
  type: "ADD_TRANSACTION",
  transaction,
});

export const startAddTransaction = (transactionData = {}) => {
  return (dispatch) => {
    const {
      currencyName = "",
      units = 0,
      amount = 0,
      createdAt = 0,
    } = transactionData;

    const transaction = {
      currencyName,
      units,
      amount,
      createdAt,
    };

    return database
      .ref("transactions")
      .push(transaction)
      .then((ref) => {
        dispatch(
          addTransaction({
            id: ref.key,
            ...transaction,
          })
        );
      });
  };
};

//Remove_Transaction

export const removeTransaction = ({ id } = {}) => ({
  type: "REMOVE_TRANSACTION",
  id,
});

export const startRemoveTransaction = ({ id }) => {
  return (dispatch) => {
    return database
      .ref(`transactions/${id}`)
      .remove()
      .then(() => {
        dispatch(removeTransaction({ id }));
      });
  };
};

// Edit_Transaction

export const editTransaction = (id, updates) => ({
  type: "EDIT_TRANSACTION",
  id,
  updates,
});

export const startEditTransaction = (id, updates) => {
  return (dispatch) => {
    return database
      .ref(`transactions/${id}`)
      .update(updates)
      .then(() => dispatch(editTransaction(id, updates)));
  };
};

export const setTransactions = (transactions) => ({
  type: "SET_TRANSACTIONS",
  transactions,
});

export const startSetTransactions = () => {
  return (dispatch) => {
    return database
      .ref("transactions")
      .once("value")
      .then((snapshot) => {
        const transactions = [];
        snapshot.forEach((childSnapshot) => {
          transactions.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setTransactions(transactions));
      });
  };
};
