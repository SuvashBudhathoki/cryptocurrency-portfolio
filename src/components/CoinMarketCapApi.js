import React from "react";
import axios from "axios";
import currencyNames from "./CurrencyNames";
import numeral from "numeral";

const apiKey = "19aff428-0470-434b-8807-cf104d68babc";

export default class CoinMarketCap extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      cryptos: []
    };
  }

  componentDidMount() {
    axios
      .get(
        "https://pro-api.coinmarketcap.com/v1/cryptocurrency/listings/latest?CMC_PRO_API_KEY=" +
          apiKey
      )
      .then(res => {
        const cryptos = res.data.data;

        this.setState(() => ({ cryptos }));
      });
  }

  handleSymbol = (crypto, currencyName) => {
    if (crypto.symbol === currencyName.symbol) {
      return (
        <div>
          {crypto.name} {crypto.symbol}
          {numeral(crypto.quote.USD.price / 0.67).format("$0,0.00")}AUD
        </div>
      );
    }
  };

  render() {
    return (
      <div>
        {this.state.cryptos.map((crypto = { id, name, symbol, quote }) =>
          currencyNames.map((currencyName = { symbol }) =>
            this.handleSymbol(crypto, currencyName)
          )
        )}
      </div>
    );
  }
}
