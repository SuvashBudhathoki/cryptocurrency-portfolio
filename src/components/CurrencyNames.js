export default [
  { value: "showAll", label: "Show All Currencies" },
  { value: "bitcoin", label: "Bitcoin", symbol: "BTC" },
  { value: "bitcoinCash", label: "Bitcoin Cash", symbol: "BCH" },
  { value: "ethereumClassic", label: "Ethereum Classic", symbol: "ETC" },
  { value: "ethereum", label: "Ethereum", symbol: "ETH" },
  { value: "litecoin", label: "Litecoin", symbol: "LTC" },
  { value: "ripple", label: "Ripple XRP", symbol: "XRP" },
  { value: "stellar", label: "Stellar", symbol: "XLM" },
  { value: "zCash", label: "Zcash", symbol: "ZEC" }
];
