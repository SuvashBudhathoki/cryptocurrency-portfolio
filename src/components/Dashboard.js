import React from "react";
import TransactionList from "./TransactionList";
import TransactionTotal from "./TransactionTotal";
import CoinMarketCap from "./CoinMarketCapApi";
import TransactionListFilters from "./TransactionListFilters";

const Dashboard = () => (
  <div>
    <p>
      <TransactionListFilters />
      <TransactionTotal />
      <TransactionList />
      ------CoinMarketCap----
    </p>
  </div>
);

export default Dashboard;
