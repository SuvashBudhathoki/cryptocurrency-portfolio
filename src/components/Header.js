import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
  <header>
    <h1>Cryptocurrency</h1>
    <NavLink to="/" exact={true}>
      Dashboard
    </NavLink>
    <NavLink to="/create">Create Transaction</NavLink>
  </header>
);

export default Header;
