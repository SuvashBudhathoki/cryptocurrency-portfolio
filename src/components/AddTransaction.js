import React from "react";
import { connect } from "react-redux";
import TransactionForm from "./TransactionForm";
import { addTransaction } from "../actions/transactions";

export class AddTransaction extends React.Component {
  onSubmit = transaction => {
    this.props.addTransaction(transaction);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Transaction</h1>{" "}
          </div>
        </div>
        <div className="content-container">
          <TransactionForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addTransaction: transaction => dispatch(addTransaction(transaction))
});

export default connect(undefined, mapDispatchToProps)(AddTransaction);
