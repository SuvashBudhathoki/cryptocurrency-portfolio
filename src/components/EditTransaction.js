import React from "react";
import { connect } from "react-redux";
import TransactionForm from "./TransactionForm";
import { editTransaction, removeTransaction } from "../actions/transactions";

export class EditTransaction extends React.Component {
  onSubmit = transaction => {
    this.props.editTransaction(this.props.transaction.id, transaction);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.removeTransaction({ id: this.props.transaction.id });
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <TransactionForm
          onSubmit={this.onSubmit}
          transaction={this.props.transaction}
        />
        <button onClick={this.onRemove}>Delete Transaction</button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  editTransaction: (id, transaction) =>
    dispatch(editTransaction(id, transaction)),
  removeTransaction: id => dispatch(removeTransaction(id))
});

const mapStateToProps = (state, props) => ({
  transaction: state.transactions.find(
    transaction => transaction.id === props.match.params.id
  )
});

export default connect(mapStateToProps, mapDispatchToProps)(EditTransaction);
