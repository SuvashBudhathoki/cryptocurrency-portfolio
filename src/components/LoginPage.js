import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";

const LoginPage = ({ startLogin }) => (
  <div class="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Cryptocurrency Portfolio</h1>
      <p>Time to manage your cryptocurrenies</p>
      <button className="button" onClick={startLogin}>
        Login with Google
      </button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
