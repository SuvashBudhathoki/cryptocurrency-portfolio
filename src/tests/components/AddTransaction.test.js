import React from "react";
import { shallow } from "enzyme";
import { AddTransaction } from "../../components/AddTransaction";
import transactions from "../fixtures/transactions";

let addTransaction, history, wrapper;

//using beforeEach function from jest to set the values for each test before executiion

beforeEach(() => {
  addTransaction = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <AddTransaction addTransaction={addTransaction} history={history} />
  );
});

test("should render AddTransaction page correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
  wrapper.find("TransactionForm").prop("onSubmit")(transactions[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(addTransaction).toHaveBeenLastCalledWith(transactions[1]);
});
