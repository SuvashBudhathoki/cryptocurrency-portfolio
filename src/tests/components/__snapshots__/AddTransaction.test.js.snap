// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render startAddTransaction page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
         Add Transaction
      </h1>
       
    </div>
  </div>
  <div
    className="content-container"
  >
    <TransactionForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
